presto-common/src/main/java/com/facebook/presto/common/block/BlockUtil.java

```diff
@@ -326,4 +326,29 @@ public static int getNum128Integers(int length)
         }
         return num128Integers;
     }
+
+    /**
+     * Returns null if all blocks are already loaded, otherwise returns a new blocks array with all blocks loaded
+     */
+    @Nullable
+    static Block[] ensureBlocksAreLoaded(Block[] blocks)
+    {
+        for (int i = 0; i < blocks.length; i++) {
+            Block loaded = blocks[i].getLoadedBlock();
+            if (loaded != blocks[i]) {
+                // Transition to new block creation mode after the first newly loaded block is encountered
+                Block[] loadedBlocks = new Block[blocks.length];
+                if (i > 0) {
+                    System.arraycopy(blocks, 0, loadedBlocks, 0, i);
```

Comment:
- noob question: this arraycopy will be shadow or deep copy? I guess shadow as we need to reuse the actual memory representation?
- Curious why we use the first `i` blocks to be newly created and the rest to be potentially not newly created. Is this for performance reasons? So if we use a full for loop from i = 0 to blocks.length on `blocks[i].getLoadedBlock()`, will that be slower than this one even all previous i blocks are loaded.

Reply:
There was some debate on the use of `System.arraycopy` vs `blocks.clone()` over in https://github.com/trinodb/trino/pull/9112#discussion_r702121568. Since we decided to stick with `blocks.clone()`, I'll try to cover the general strategy and less the specifics around `System.arraycopy`.

> this ... will be shadow or deep copy? I guess shadow as we need to reuse the actual memory representation?

Yes, when `blocks[i].getLoadedBlock() == blocks[i]` we can just copy the references between the arrays directly without re-invoking `Block#getLoadedBlock()` on the elements which we already know do not change.

> Curious why we use the first `i` blocks to be newly created and the rest to be potentially not newly created. Is this for performance reasons? So if we use a full for loop from i = 0 to blocks.length on blocks[i].getLoadedBlock(), will that be slower than this one even all previous i blocks are loaded.

Yeah, using the references that we already know don't change is desirable in general, because `Block#getLoadedBlock()` is almost certainly megamorphic at this (and similar) call sites, so in general avoiding the redundant calls is worthwhile. When the blocks array is small, this difference is negligible, but when the blocks arrays are larger and the first changed block position is towards the middle or end of the array then the difference can be significant.